src/vite-env.d.ts

```diff
@@ -11,8 +21,18 @@ declare global {
   const bridgeVersion: string;
 }
 
+// Module augmentation does not allow changing the types of existing properties (https://github.com/microsoft/TypeScript/issues/36146)
+// but the types of Ethers v5 are broken (https://github.com/ethers-io/ethers.js/issues/2325#issuecomment-977620740), so we provide here
+// a typesafe version of the types and functions that we use in this project
+
+declare module "@ethersproject/abstract-provider" {
+  type TypeSafeTransactionResponse = Nullable<TransactionResponse>;
```

Comment:
Mmmm why setting all the properties of `TransactionResponse` and `TransactionReceipt` to `type | null`? This is not true, only some properties can be `null`.

Reply:
Yep, but we don't know which ones. The types of Ethers v5 are broken and Richard Moore, the main maintainer of the project, [recommends we assume that any property can be `null`](https://github.com/ethers-io/ethers.js/issues/2325#issuecomment-977620740):

<img src="https://user-images.githubusercontent.com/15896177/215547814-4b7eb8d1-baf7-4bc6-81e4-2a9a30e861ee.png" width="400px" />

I added this because I (once again) ran into a runtime error. We've fallen into this trap multiple times, so it's best to be safe and (as they recommend) always assume that all properties can be null. It's always better to extra null check than to get runtime errors.

Reply:
Well, he says that because how the typings have being built. But he also says:

> ... any function needs to be assumed to be null *(although, most of them won’t be)*.

Rewriting the types of all the functions that we use from the library is not an option, even for being type safe. Doing this overrides we can get more runtime errors if at some point they decide to return anything different from the functions. I can accept this changes as an exception because I know that they can return `null` and they are important, but we should not override more ethers types as this have more drawbacks than a runtime error.

Reply:
> Rewriting the types of all the functions that we use from the library is not an option, even for being type safe.

Of course not! We don't need to override the utils for example. I think this only applies to the values returned by the functions that interact with the blockchain. At least there is where I've found this problem.

> Doing this overrides we can get more runtime errors if at some point they decide to return anything different from the functions.

I don't think this is true and if it were it would not be acceptable. They can return whatever they want in future versions and we won't get any runtime errors. The mapped type `Nullable<Type>` only extends the type they declare (whatever it's today or in any future version) to include `null`, so we're forced to null check. I don't see how this can lead to future runtime errors. As I understand it, this only forces us to be more verbose for the benefit of type safety.

Reply:
> The mapped type Nullable<Type> only extends the type they declare *(whatever it's today or in any future version)* to include null, so we're forced to null check.

This is not true, as they can change the interface being returned by the function. Overriding this functions, you won't notice if the interface being returned changes or not. You will only notice it if the import fails (the interface returned is removed from the library). But it's very common to see things like `TransactionResponseV2` when transitioning from one version of a library to another. In this scenario, if they keep the `TransactionResponse` as it's right now, you are doing a wrong assumption.

This can be improved using the [`ReturnType`](https://www.typescriptlang.org/docs/handbook/utility-types.html#returntypetype) util BTW.

Reply:
I think we're safe in the event of the interface not being available (name change) or having a structural change since in these cases we would get an error at compile time:

<img src="https://user-images.githubusercontent.com/15896177/216009762-8c95ec2d-80c3-4779-8d32-c660b9def432.gif" width="400px" />

It's true though that if the function disappears or is renamed, our augmentation would produce a runtime error, but what are the odds of this happening?

While it's technically possible to have a runtime error if the function disappears, I don't think it's a realistic scenario that we have to consider in the real world, precisely because library authors are reluctant to make changes that break backward compatibility.